tests/PHPStan/Type/Constant/ConstantArrayTypeBuilderTest.php

```diff
@@ -128,4 +129,29 @@ public function testIsList(): void
 		$this->assertFalse($builder->isList());
 	}
 
+	public function testIsListWithUnion(): void
+	{
+		$builder = ConstantArrayTypeBuilder::createEmpty();
+
+		$builder->setOffsetValueType(null, new ConstantIntegerType(0));
+		$this->assertTrue($builder->isList());
+
+		$builder->setOffsetValueType(new ConstantIntegerType(0), new NullType());
+		$this->assertTrue($builder->isList());
+
+		$builder->setOffsetValueType(new ConstantIntegerType(1), new NullType());
+		$this->assertTrue($builder->isList());
+
+		$builder->setOffsetValueType(new ConstantIntegerType(2), new NullType());
+		$this->assertTrue($builder->isList());
+
+		$oneOrZero = TypeCombinator::union(
```

Comment:
Please also write some counter-example test as well. Like when we're setting a union offset that does not exist yet, that will make the array be no longer a list. Thanks.